godot-core/src/engine/translate.rs

```diff
@@ -0,0 +1,110 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+//! Macros for translation.
+
+pub use crate::{tr, tr_n};
+
+/// A convenience macro for using the [`Object::tr()`](crate::engine::Object::tr()) and [`Object::tr_ex()`](crate::engine::Object::tr_ex())
+///  methods.
+///
+/// The methods are called from the [`Engine`](crate::engine::Engine) singleton.
+///
+/// Takes a format string literal, with optional arguments. Optionally, `context` for potentially ambiguous words can be
+/// added before the format arguments, separated with a `;`.
+///
+/// For maximum flexibility and minimum ambiguity, prefer inlined, positional, or named arguments in the format strings:
+/// ```no_run
+/// # use godot_core::tr;
+/// # struct Foo { x: i32, y: i32 };
+/// # impl std::fmt::Display for Foo {
+/// #    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+/// #        write!(f, "")
+/// #    }
+/// # }
+/// # let a = Foo { x: 0, y: 0 };
+/// # let b = Foo { x: 0, y: 0 };
+/// # let context = "context";
+/// // Good
+/// tr!(context; "{a} is a {b}"); // inlined, with context
+/// tr!("{0} is a {1}", a, b); // positional, without context
+/// tr!("{c} is a {d}", c = a.x, d = b.y); // named (inlining not possible here)
+///
+/// // Not so good, much more fragile
+/// // and does not easily allow ordering arguments differently in other translations.
+/// tr!("{} is a {}", a, b);
+/// ```
+/// see also: <https://docs.godotengine.org/en/stable/tutorials/i18n/internationalizing_games.html#translation-contexts>
+#[macro_export]
+macro_rules! tr {
+    ($fmt:literal $(, $($args:tt)*)?) => {
+        $crate::engine::Engine::singleton()
+            .tr(format!($fmt $(, $($args)*)?).into())
+    };
+
+    ($cxt:expr; $fmt:literal $(, $($args:tt)*)?) => {
+        $crate::engine::Engine::singleton()
+            .tr_ex(format!($fmt $(, $($args)*)?).into())
+            .context(format!("{}", $cxt).into())
+            .done()
+    };
+}
+
+/// A convenience macro for using the [`Object::tr_n()`](crate::engine::Object::tr_n()) and
+/// [`Object::tr_n_ex()`](crate::engine::Object::tr_n_ex()) methods.
+///
+/// The methods are called from the [`Engine`](crate::engine::Engine) singleton.
+///
+/// Allows for the use of format strings with arbitrary arguments. `n` is given prior to the format string, followed by `;`.
+/// Optionally, `context` for potentially ambiguous words can be added with `,` after `n` and before `;`.
+///
+/// For maximum flexibility and minimum ambiguity, prefer inlined, positional, or named arguments in the format strings:
+/// ```no_run
+/// # use godot_core::tr_n;
+/// # struct Foo { x: i32, y: i32 };
+/// # impl std::fmt::Display for Foo {
+/// #    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+/// #        write!(f, "")
+/// #    }
+/// # }
+/// # let a = Foo { x: 0, y: 0 };
+/// # let b = Foo { x: 0, y: 0 };
```

Comment:
`godot_core` is not a public crate, so the import could use `godot` instead. We could even show this one, so that users know what to import.

You can also use `Vector2i` instead of defining your own `Foo`:
```suggestion
/// # use godot::builtin::Vector2i;
/// # let a = Vector2i::ZERO;
/// # let b = Vector2i::ZERO;
use godot::engine::translate::tr_n;

```
(empty line after visible import)

Reply:
In case this creates problems with "module godot not found", feel free to use the "godot_core" approach (should not happen though...)

Reply:
It *was* creating issues originally, Once I got the `#[macro_use]` tag figured out it was smooth sailing.

Reply:
Interesting, I've never had to do that 🙂 but if it works, great!